planning/MartenV4-WastazComments.md

```diff
@@ -0,0 +1,174 @@
+you probably need to mark a ton of shit as nullable in every document/event you have.
+
+We have tried both using optimistic and pessimistic models. Pessimistic models leads to...code that you dont really want. But optimistic models however leads to either a false sense of
+security or you have to force a validation step for your document after each load. Thing that we think works best so far has been building validators with FluentValidation for every event/document
+type, using optimistic models and validating the documents after reading them. Its...not that fun, but it works and it lets us actually model the documents/events according to how they should look
+according to the domain.
+
+However there are some ways to make this easier.
+Ive seen the discussions on the "streaming api", but I think that what may be missing from these discussions is that there can actually be tremendous alternate value in providing a way to
+easily read raw documents/events as `string` or `Stream`. Because it allows an easier way of hooking in a _parser_, not a _serializer_. There's some good blog posts out there on the virtues of
+parsing instead of validating (or serializing) so Im not gonna repeat the points that they already made. But with this type of streaming api you could much easier hook up a parser which would not
+only help in having an easier way of constructing correct objects that respects NRTs, but could also help a lot with the F# side of this where immutability and discriminated unions are not always the
+best at serializing/deserializing. You can probably do some things like that with the `ISerializer` interface today, but having a way to just bypass serialization completely is quite powerful.
+
+# Event versioning
+
+Event versioning is hard. Greg Youngs book on the topic almost does justice to how hard it is. Its hard. Marten doesnt really have any support for event versioning out of the box, apart from the
+usual "serializer can handle putting default values on added fields". But it would certainly help to have more powerful choices here in upgrading events as they evolve. I dont have a great suggestion
+on exactly how this should be done though, and sadly I dont really love any solution that I have seen in other libs/frameworks attempting to do this either. Currently I tend to write a migration reading
+the events, parsing them as `JObject` or `JToken` and working with the json directly and then updating. It works. Its not great, but it works. There's certainly space here for thinking about this should be done.
+
+# Snapshots
+
+Snapshots are useful. The main input I have here is that _dont fall into the same trap as a lot of other ES libraries on the .NET platform_. And the trap that I am talking about is the assumption
+that each stream should have one and only one singe aggregate type associated with it. Marten can currently write a stream type in `mt_streams` which sadly is generated from a .NET type name, Id
```

Comment:
We don't plan to make a type mandatory, but it might be used for some other cases. Currently it's possible to use string as the id (however I think that we have some bug on that).

As I wrote in another comment - we should make snapshot generation more flexible (to have possibility to decide when it's happening and store more than one snapshot of same stream).